rules/aip0122/resource_name_components_alternate.go

```diff
@@ -0,0 +1,57 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aip0122
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/googleapis/api-linter/lint"
+	"github.com/googleapis/api-linter/locations"
+	"github.com/googleapis/api-linter/rules/internal/utils"
+	"github.com/jhump/protoreflect/desc"
+)
+
+var resourceNameComponentsAlternate = &lint.MessageRule{
+	Name:   lint.NewRuleName(122, "resource-name-components-alternate"),
+	OnlyIf: utils.IsResource,
+	LintMessage: func(m *desc.MessageDescriptor) []lint.Problem {
+		var problems []lint.Problem
+		resource := utils.GetResource(m)
+		for _, p := range resource.GetPattern() {
+			expectIdentifier := false
+			components := strings.Split(p, "/")
+			for _, c := range components {
+				if expectIdentifier != isIdentifier(c) {
+					problems = append(problems, lint.Problem{
+						Message: fmt.Sprintf(
+							"Resource pattern %q must alternate between collection and identifier.",
+							p,
+						),
+						Descriptor: m,
+						Location:   locations.MessageResource(m),
+					})
+					break
+				}
+				expectIdentifier = !expectIdentifier
+			}
```

Comment:
```suggestion
			parts := strings.Split(p, "/")
			for i, part := range parts {
				if i % 2 == 1 && !isIdentifier(c) {
					problems = append(problems, lint.Problem{
						Message: fmt.Sprintf("Resource pattern %q must alternate between collection and identifier. %q is not an identifier", p, part),
						Descriptor: m,
						Location:   locations.MessageResource(m),
					})

					break
				}
			}
```

I find the switching `expectIdentifier` to be confusing. Use of the mod operator seems clearer?

Reply:
Also, go has no line length restrictions, and putting the whole of fmt.Sprintf on one line would be more common.

Reply:
ack on both, I applied the suggestion but left "parts" as "components", since
that's the verbage used in the AIP: https://google.aip.dev/122#guidance.